### DISCUSSION TITLE
How to update a mention once it got renamed or deleted?

### DISCUSSION THREAD

**Discussion by tomaszferens** [facebook/lexical#2646] [Q&A]
I want to build a mention feature in the editor. I started by copying the `MentionsPlugin`, `MentionNode` and `LexicalTypeaheadMenuPlugin` from the playground.

In the playground, when you use a mention and export state into JSON (that's the format I want to store in the database) it outputs something like this for `MentionNode`:

```ts
{
  detail: 1,
  mentionName: "Adi Gallia",
  format: 0,
  mode: "segmented",
  style: "",
  text: "Adi Gallia",
  type: "mention",
  version: 1,
}
```

The problem is, that it doesn't store the `id` of the element, so I don't know what has got mentioned. I changed the schema of `MentionNode` to return something like this:

```ts
{
  detail: 1,
  entity: "user",
  entityId: "test-id",
  format: 0,
  mode: "segmented",
  style: "",
  text: "Adi Gallia",
  type: "mention",
  version: 1,
};
```

I removed `mentionName` and added `entity` and `entityId` so that I can later on react if something changes. 

Now, what if user (or whatever other object we mention) renames? We have the `id` so can probably update the schema, but where to do it - frontend or backend?

I think the easiest thing is to traverse the saved state on the backend, collect all mentioned `ids` and get the fresh mentions? 🤔 How to iterate over the lexical state - is there some helper around that?

**Comment by CanRau** [comment]
hey, quite old though I'm curious did you figure it out?

I think it depends on your requirements, but one thing I started doing in my custom `OnChangePlugin` is not only "surfacing" the editor state but also `details` which is an object that contains for example all mentions, headings and more which I populate like so 
```ts
details.mentions = lexical.$nodesOfType(MentionNode).map((n) => n.exportJSON());
```

which I then store separate from the tree, currently all in 1 json column but possibly going to store mentions for example in a separate table so I can store more details like if that user has been notified etc and also have easy back-links.

Not sure about the need to store an id in a MentionNode, though you'd need to keep track of which mentions are still in the current lexical state and which ones have been removed.

**Reply by CanRau** [reply]
I mean yea it's actually interesting to maybe even only store a mention id in the tree and have the rest handled separately so you'd join all mentioned user details in or sthg, this way the mentions don't care about the mentioned username etc so when the user updates you don't need to traverse all lexical state trees and update all mentions.
Just remembered that I also still need to do this actually 😅
